feat(keyEvents): support for <div (keyup.enter)="callback()">

divdavem opened the pull request:
This pull request is a first implementation of #523. It adds a plugin for the event manager, to allow a key name to be appended to the event name (for keyup and keydown events), so that the callback is only called for that key.

Here are some examples:
 (keydown.shift.enter)
 (keyup.space)
 (keydown.control.shift.a)
 (keyup.f1)

Key names mostly follow the DOM Level 3 event key values:
http://www.w3.org/TR/DOM-Level-3-Events-key/#key-value-tables

There are some limitations to be worked on (see later), but for now, this implementation is reliable for the following keys (by "reliable" I mean compatible with Chrome and Firefox and not depending on the keyboard layout):
- alt, control, shift, meta (those keys can be combined with other keys)
- tab, enter, backspace, pause, scrolllock, capslock, numlock
- insert, delete, home, end, pageup, pagedown
- arrowup, arrowdown, arrowleft, arrowright
- latin letters (a-z), function keys (f1-f12)
- numbers on the numeric keypad (but those keys are not correctly simulated by Chromedriver)

There is a sample to play with in examples/src/key_events/.

Here is a list of limitations/items to work on:
- Chrome and Firefox have different behavior, because they implement different versions of the DOM level 3 events specification (Chrome implements an old draft version with the "keyIdentifier" property, Firefox implements the latest draft version with the "key" property).
- Especially, using the shift modifier changes the value of the "key" property, but not the value of the "keyIdentifier" property, which means "shift.1" on Chrome is reported as "shift:!" on Firefox (for the US qwerty keyboard).
- As a consequence, on Firefox, keys can quickly be linked with a keyboard layout, for example: "shift.:" only works on layouts for which it is needed to press shift to get the colon (which is the case on US qwerty keyboards bot not on French azerty keyboards).
- Chrome reports strange values for some keys (most keys not reported in my previous list, such as: ":", ";", "?", "<", ">" ...).
- Chromedriver does not correctly simulate keys from the numeric keypad
- It would probably be a good idea to add support for a "command" or "accel" modifier which would be "meta" on Mac and "control" on other systems (cf [here](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/getModifierState#.22Accel.22_virtual_modifier))

vicb commented on modules/angular2/src/core/events/key_events.js:
```diff
@@ -0,0 +1,171 @@
+
+  static parseEventName(eventName: string) {
+    eventName = eventName.toLowerCase();
+    var parts = StringWrapper.split(eventName, splitRegExp);
```
eventName.split('.') & drop the regexp

vicb commented on modules/angular2/src/core/events/key_events.js:
```diff
@@ -0,0 +1,171 @@
+import {StringMapWrapper, ListWrapper} from 'angular2/src/facade/collection';
+import {EventManagerPlugin} from './event_manager';
+
+var splitRegExp = RegExpWrapper.create("\\.");
```
could you please use double or single quotes consistently (single preferred)

vicb commented on modules/angular2/src/core/events/key_events.js:
```diff
@@ -0,0 +1,171 @@
+  }
+
+  supports(eventName: string): boolean {
+    return !! KeyEventsPlugin.parseEventName(eventName);
```
!! would not work with Dart, isPresent() ?

vicb commented on modules/angular2/src/core/events/key_events.js:
```diff
@@ -0,0 +1,171 @@
+    var fullKey = "";
+    ListWrapper.forEach(modifierKeys, (modifierName) => {
+      if (ListWrapper.contains(parts, modifierName)) {
+        ListWrapper.remove(parts, modifierName);
```
Not sure if Dart allow removing in a forEach. 
Anyway a counter might perform better.

vicb commented on modules/angular2/src/core/events/key_events.js:
```diff
@@ -0,0 +1,171 @@
+    eventName = eventName.toLowerCase();
+    var parts = StringWrapper.split(eventName, splitRegExp);
+    var domEventName = ListWrapper.removeAt(parts,0);
+    if ( (parts.length == 0) || (domEventName !== "keydown" && domEventName !== "keyup")) {
```
minor: 
- WS after , in `ListWrapper.removeAt(parts,0);`
- no WS  after `if (`

vicb commented on modules/angular2/src/core/events/key_events.js:
```diff
@@ -0,0 +1,171 @@
+      // keyIdentifier is defined in the old draft of DOM Level 3 Events implemented by Chrome and Safari
+      // cf http://www.w3.org/TR/2007/WD-DOM-Level-3-Events-20071221/events.html#Events-KeyboardEvents-Interfaces
+      if (isPresent(key) && StringWrapper.startsWith(key,"U+")) {
+        key = StringWrapper.fromCharCode(NumberWrapper.parseInt(StringWrapper.substring(key,2),16));
```
, + WS

vicb commented on modules/angular2/src/core/events/key_events.js:
```diff
@@ -0,0 +1,171 @@
+      // cf http://www.w3.org/TR/2007/WD-DOM-Level-3-Events-20071221/events.html#Events-KeyboardEvents-Interfaces
+      if (isPresent(key) && StringWrapper.startsWith(key,"U+")) {
+        key = StringWrapper.fromCharCode(NumberWrapper.parseInt(StringWrapper.substring(key,2),16));
+        if (event.location == 3 /* numeric keypad */ && StringMapWrapper.contains(chromeNumKeyPadMap, key)) {
```
3 -> constant ?

vicb commented on modules/angular2/src/core/events/key_events.js:
```diff
@@ -0,0 +1,171 @@
+    };
+  }
+
+  static getEventKey(event) {
```
add return type :string

vicb commented on modules/angular2/src/core/events/key_events.js:
```diff
@@ -0,0 +1,171 @@
+    return key;
+  }
+
+  static getEventFullKey(event) {
```
return type

vicb commented on modules/angular2/src/core/events/key_events.js:
```diff
@@ -0,0 +1,170 @@
+
+    var outsideHandler = shouldSupportBubble ?
+      KeyEventsPlugin.bubbleCallback(element, parsedEvent.fullKey, handler, this.manager.getZone()) :
+      KeyEventsPlugin.sameElementCallback(element, parsedEvent.fullKey, handler, this.manager.getZone());
```
factorize the 2 methods by passing `shouldSupportBubble` ?

vicb commented on modules/examples/src/key_events/index.js:
```diff
@@ -0,0 +1,30 @@
+}
+
+export function main() {
+  bootstrap(KeyEventsApp);
```
you need to initialize the reflector for this to work in Dart, see other examples

vicb commented on modules/examples/src/key_events/index.js:
```diff
@@ -0,0 +1,30 @@
+  }
+
+  onShiftEnter(event) {
+    alert('You pressed shift.enter!');
```
alert would not work in Dart

vicb commented on modules/angular2/src/core/events/key_events.js:
```diff
@@ -0,0 +1,170 @@
+    }
+    if (key.length == 0) {
+      throw new BaseException('A key must be specified.');
+    }
```
not sure if we should throw here or only return `null`

Someone might with a plugin that would support `keydown.fancymodifier` chained with this one

vicb commented on modules/angular2/src/core/events/key_events.js:
```diff
@@ -0,0 +1,170 @@
+  'N':'.',
+  'O':'/',
+  '\x60': '0',
+  '\x90':'numlock'
```
nitpick : + WS (consistency)

divdavem commented:
@vicb Thank you very much for reviewing my pull request so quickly, and thank you for your comments. I have done all the changes you suggested.
The travis-ci failure for the js part does not seem to be linked with my changes. However, I can see [here](https://travis-ci.org/angular/angular/jobs/56131235#L2092) that the e2e test I added is still failing for dart:

> http://localhost:8002/examples/src/key_events/index.dart.js 1791:3 Uncaught Uncaught Error: NoSuchMethodError: undefined is not a function

Unfortunately, dart does not work well on my Windows machine currently, so I am unable to check what's going wrong right now, but I will try to fix it.
For info, I will probably not have time to look at it before Monday.
Have a good week-end!

vicb commented on modules/angular2/src/core/events/key_events.js:
```diff
@@ -0,0 +1,162 @@
+  }
+
+  static getEventKey(event): string {
+    var key = event.key;
```
`key` does not seem to be a valid Dart property, see https://api.dartlang.org/apidocs/channels/stable/dartdoc-viewer/dart:html.KeyboardEvent

divdavem replied:
@vicb Thank you for you comment.

As far as I can see, there is no equivalent for the [standard `key` property](http://www.w3.org/TR/uievents/#widl-KeyboardEvent-key). The `keyCode` property could be used, but it is [deprecated](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode):

> This shouldn't be used by new web applications. IE and Firefox already (partially) support KeyboardEvent.key. Also, Google Chrome and Safari support KeyboardEvent.keyIdentifier which is defined in the old draft of DOM Level 3 Events. If you can use them or one of them, you should use them/it as far as possible.

Is there any hope that Dart can be updated soon to follow web standards regarding key events?
Or should we include in the Dart version of angular2 a map of `keyCode` values to their corresponding `key` values? I feel that it does not make much sense, because the Dart code is ultimately transpiled to JS (as [Dart will not be supported natively in Chrome](http://news.dartlang.org/2015/03/dart-for-entire-web.html)), and the `key` or `keyIdentifier` property would be available if the check done by `Dart` was not so strict... Is there any way to bypass the strict check from Dart?

What do you suggest?

Also, for info, I have tried today to make the dart build work on my Windows machine, but it is still failing, both on my branch and on master. Here is the error I am getting when running `gulp build.dart`:

> dist\dart\angular2\:[hint] Unused import (D:\git_clones\angular\dist\dart\angular2\test\core\compiler\view_container_spec.dart, line 8, col 8)
> dist\dart\angular2\:[hint] Unused import (D:\git_clones\angular\dist\dart\angular2\test\core\compiler\view_spec.dart, line 20, col 8)
> dist\dart\angular2\:[hint] Unused import (D:\git_clones\angular\dist\dart\angular2\test\test_lib\test_lib_spec.dart, line 16, col 8)
> [16:37:52] 'build/analyze.dart' errored after 8.06 s
> [16:37:52] Error: Dartanalyzer showed hints
>    at ChildProcess.<anonymous> (D:\git_clones\angular\tools\build\dartanalyzer.js:84:19)
>    at ChildProcess.emit (events.js:98:17)
>    at maybeClose (child_process.js:756:16)
>    at Process.ChildProcess._handle.onexit (child_process.js:823:5)
> [16:37:52] 'build.dart' errored after 53 s
> [16:37:52] Error: [object Object]
>    at formatError (D:\git_clones\nodist\bin\node_modules\gulp\bin\gulp.js:169:10)
>    at Gulp.<anonymous> (D:\git_clones\nodist\bin\node_modules\gulp\bin\gulp.js:195:15)
>    at Gulp.emit (events.js:117:20)
>    at Gulp.Orchestrator._emitTaskDone (D:\git_clones\angular\node_modules\gulp\node_modules\orchestrator\index.js:264:8)
>    at D:\git_clones\angular\node_modules\gulp\node_modules\orchestrator\index.js:275:23
>    at finish (D:\git_clones\angular\node_modules\gulp\node_modules\orchestrator\lib\runTask.js:21:8)
>    at cb (D:\git_clones\angular\node_modules\gulp\node_modules\orchestrator\lib\runTask.js:29:3)
>    at finish (D:\git_clones\angular\node_modules\run-sequence\index.js:38:5)
>    at Gulp.onError (D:\git_clones\angular\node_modules\run-sequence\index.js:45:4)
>    at Gulp.emit (events.js:117:20)

vicb commented:
@divdavem you should probably check dartbug.com or create a new issue there to get some information/plan about implementation plans.
In the meantime, Dart should also be supported, should you create a separate .dart file with a map.

Your build error seems to be caused by unused imports
> dist\dart\angular2:[hint] Unused import (D:\git_clones\angular\dist\dart\angular2\test\core\compiler\view_container_spec.dart, line 8, col 8)
dist\dart\angular2:[hint] Unused import (D:\git_clones\angular\dist\dart\angular2\test\core\compiler\view_spec.dart, line 20, col 8)
dist\dart\angular2:[hint] Unused import (D:\git_clones\angular\dist\dart\angular2\test\test_lib\test_lib_spec.dart, line 16, col 8)

The strange thing is that a first look your PR does not seem to affect those files, can you build the master branch ?

vicb commented:
@divdavem what is the status on this PR ?

divdavem commented:
@vicb Thank you for your comments.
I have updated the PR to make it compatible with dart (I used a virtual machine with Linux, as I gave up trying to make dart work on my Windows machine) and I added unit tests for the `parseEventName` function (to follow the suggestion from @pkozlowski-opensource ).
Note that the map to convert `keyCode` values to `key` values for Dart does not include all possible keys (only those I reported as reliable at the beginning of this PR). The missing keys will be reported as `'Unidentified'`.
Could you have a look at my PR again? It could be merged if it is ok.

vicb commented on modules/angular2/src/dom/browser_adapter.dart:
```diff
@@ -14,6 +14,87 @@ class _IdentitySanitizer implements NodeTreeSanitizer {
 final _identitySanitizer = new _IdentitySanitizer();
 
+final _keyCodeToKeyMap = const {
+  8: 'Backspace',
```
I'm wondering if we should add constants for "special" keys, ie
`var KEY_BACKSPACE = 'Backspace';`

vicb commented on modules/angular2/src/dom/browser_adapter.es6:
```diff
@@ -286,4 +329,27 @@ export class BrowserDomAdapter extends GenericBrowserDomAdapter {
+      // keyIdentifier is defined in the old draft of DOM Level 3 Events implemented by Chrome and Safari
+      // cf http://www.w3.org/TR/2007/WD-DOM-Level-3-Events-20071221/events.html#Events-KeyboardEvents-Interfaces
+      if (isBlank(key)) {
+        key = 'Unidentified';
```
`return 'Unidentified';` and remove the "`else`" ?
I don't think it makes sense to `_keyMap.hasOwnProperty(key)` in this case ?

vicb commented on modules/angular2/src/render/dom/events/key_events.js:
```diff
@@ -0,0 +1,93 @@
+    });
+  }
+
+  static parseEventName(eventName: string) {
```
add an inline comment, structure of returned object

vicb commented:
lgtm with a few minor comments

divdavem committed:
feat(keyEvents): support for <div (keyup.enter)="callback()">

This commit adds a plugin for the event manager, to allow a key name to
be appended to the event name (for keyup and keydown events), so that
the callback is only called for that key.

Here are some examples:
 (keydown.shift.enter)
 (keyup.space)
 (keydown.control.shift.a)
 (keyup.f1)

Key names mostly follow the DOM Level 3 event key values:
http://www.w3.org/TR/DOM-Level-3-Events-key/#key-value-tables

There are some limitations to be worked on (cf details
in https://github.com/angular/angular/pull/1136) but for now, this
implementation is reliable for the following keys (by "reliable" I mean
compatible with Chrome and Firefox and not depending on the keyboard
layout):
- alt, control, shift, meta (those keys can be combined with other keys)
- tab, enter, backspace, pause, scrolllock, capslock, numlock
- insert, delete, home, end, pageup, pagedown
- arrowup, arrowdown, arrowleft, arrowright
- latin letters (a-z), function keys (f1-f12)
- numbers on the numeric keypad (but those keys are not correctly simulated
by Chromedriver)

There is a sample to play with in examples/src/key_events/.

close #523
close #1136

divdavem commented:
@vicb Thank you for your review. I have updated the code.
